File: electron/dataHandle/metadata.ts
import userData from "./userData";

class _MetaData {
  private readonly file = "metadata.json"
  private parse(raw: string | null): Record<string, unknown> {
    if (!raw) return {};
    try { return JSON.parse(raw); } catch { return {}; }
  }

  async read(): Promise<Record<string, unknown>>;
  async read<T = unknown>(key: string): Promise<T | null>;
  async read<T = unknown>(key?: string): Promise<Record<string, unknown> | T | null> {
    const raw = await userData.read<string>(this.file);
    const data = this.parse(raw);
    if (key === undefined) return data;
    return (data[key] as T) ?? null;
  }

  async write(data: Record<string, unknown>): Promise<boolean> {
    try {
      await userData.write(this.file, data);
      return true;
    } catch (error) {
      console.error("[metadata] Failed to write:", error);
      return false;
    }
  }

  async update(patch: Record<string, unknown>): Promise<boolean> {
    const current = await this.read();
    return this.write({ ...current, ...patch });
  }
};

export default new _MetaData();

File: src-tauri/src/service/meta_data.rs
use super::user_data::UserData;
use serde_json::{Map, Value};
use tauri::AppHandle;

pub struct MetaData {
    ud: UserData,
    file: &'static str,
}

impl MetaData {
    pub fn new(app: &AppHandle) -> Self {
        Self {
            ud: UserData::new(app, ""),
            file: "metadata.json",
        }
    }

    /// Đọc toàn bộ file → trả về Map (tương đương Record<string, unknown>)
    pub async fn read_all(&self) -> Map<String, Value> {
        self.ud
            .read::<Map<String, Value>>(self.file, None)
            .await
            .unwrap_or_default() // file không tồn tại → {}
    }

    /// Đọc theo key → trả về Option<T> (tương đương T | null)
    pub async fn read<T>(&self, key: &str) -> Option<T>
    where
        T: serde::de::DeserializeOwned,
    {
        let data = self.read_all().await;
        let value = data.get(key)?; // key không tồn tại → None
        serde_json::from_value(value.clone()).ok() // deserialize thất bại → None
    }

    /// Ghi đè toàn bộ file
    pub async fn write(&self, data: &Map<String, Value>) -> bool {
        match self.ud.write(self.file, data).await {
            Ok(_) => true,
            Err(e) => {
                eprintln!("[metadata] Failed to write: {}", e);
                false
            }
        }
    }

    /// Merge patch vào data hiện tại (tương đương spread { ...current, ...patch })
    pub async fn update(&self, patch: Map<String, Value>) -> bool {
        let mut current = self.read_all().await;
        current.extend(patch); // key trùng → patch thắng, giống JS spread
        self.write(&current).await
    }

    /// Xoá một key khỏi metadata
    pub async fn delete(&self, key: &str) -> bool {
        let mut current = self.read_all().await;
        current.remove(key);
        self.write(&current).await
    }
}